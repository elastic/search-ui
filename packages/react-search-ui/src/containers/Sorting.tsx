import React, { Component } from "react";
import { withSearch } from "..";
import {
  Sorting,
  SortingContainerContext,
  SortingContainerProps,
  SortingViewProps
} from "@elastic/react-search-ui-views";

function findSortOption(sortOptions, sortData) {
  if (sortData.indexOf("|||") === -1) {
    return sortOptions.find(
      (option) => JSON.stringify(option.value) === sortData
    );
  }

  const [value, direction] = sortData.split("|||");

  return sortOptions.find(
    (option) => option.value === value && option.direction === direction
  );
}

function formatValue(sortField, sortDirection, sortList) {
  if (sortList && sortList.length > 0) {
    return JSON.stringify(sortList);
  }
  return `${sortField}|||${sortDirection}`;
}

function formatSelectValues(sortOption) {
  if (Array.isArray(sortOption.value)) {
    // save value as string for comparison
    return JSON.stringify(sortOption.value);
  } else {
    return formatValue(sortOption.value, sortOption.direction, null);
  }
}

function formatSelectOption(sortOption) {
  return {
    label: sortOption.name,
    value: formatSelectValues(sortOption)
  };
}

export class SortingContainer extends Component<SortingContainerProps> {
  render() {
    const {
      className,
      label,
      setSort,
      sortDirection,
      sortField,
      sortList,
      sortOptions,
      view,
      ...rest
    } = this.props;

    const View = view || Sorting;

    const viewProps: SortingViewProps = {
      className,
      label,
      onChange: (o) => {
        const sortOption = findSortOption(sortOptions, o);
        setSort(sortOption.value, sortOption.direction);
      },
      options: sortOptions.map(formatSelectOption),
      value: formatValue(sortField, sortDirection, sortList),
      ...rest
    };

    return <View {...viewProps} />;
  }
}

export default withSearch<SortingContainerProps, SortingContainerContext>(
  ({ sortDirection, sortField, sortList, setSort }) => ({
    sortDirection,
    sortField,
    sortList,
    setSort
  })
)(SortingContainer);
