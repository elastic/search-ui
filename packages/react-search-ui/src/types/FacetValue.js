import PropTypes from "prop-types";
import FilterValue from "./FilterValue";

export default PropTypes.shape({
  // Number of results for this filter
  count: PropTypes.number.isRequired,
  // Filter to apply if selected
  value: FilterValue.isRequired,
  // Whether or not this facet value is selected
  selected: PropTypes.bool
});
