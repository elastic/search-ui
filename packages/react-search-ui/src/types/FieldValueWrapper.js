import PropTypes from "prop-types";

import FieldValue from "./FieldValue";

export default PropTypes.shape({
  // A raw field value, like 'I am a raw result', or 2, or true. Raw values may
  // or may not be html escaped, so *always* sanitize a raw value before rendering
  // it on a page as html.
  raw: FieldValue,
  // A snippet value contains a highlighted value. I.e., 'I <em>am</em> a raw
  // result'. These are always sanitized and safe to render as html.
  snippet: PropTypes.string
});

export function isFieldValueWrapper(object) {
  return (
    object && (object.hasOwnProperty("raw") || object.hasOwnProperty("snippet"))
  );
}
