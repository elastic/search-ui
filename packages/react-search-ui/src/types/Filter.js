import PropTypes from "prop-types";
import FilterType from "./FilterType";
import FilterValue from "./FilterValue";

export default PropTypes.shape({
  field: PropTypes.string.isRequired,
  values: PropTypes.arrayOf(FilterValue).isRequired,
  type: FilterType.isRequired
});
