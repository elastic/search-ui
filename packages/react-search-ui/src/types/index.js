export { default as Date } from "./Date";
export { default as Facet } from "./Facet";
export { default as FacetType } from "./FacetType";
export { default as FacetValue } from "./FacetValue";
export { default as Filter } from "./Filter";
export { default as FilterType } from "./FilterType";
export { default as FilterValue } from "./FilterValue";
export { default as FilterValueValue } from "./FilterValueValue";
export { default as FilterValueRange } from "./FilterValueRange";
export { default as Geo } from "./Geo";
export { default as Number } from "./Number";
export { default as Text } from "./Text";

export { default as SortOption } from "./SortOption";
export { default as Result } from "./Result";
