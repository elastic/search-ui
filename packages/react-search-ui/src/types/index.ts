import type { Facet, Filter } from "@elastic/search-ui";

export type BaseContainerProps = {
  children?: React.ReactNode;
  className?: string;
};

export type BaseContainerStateProps = {
  error: string;
  filters: Filter[];
  facets: Record<string, Facet>;
};

// From SO https://stackoverflow.com/a/59071783
// TS Utility to rename keys in a type
// Usage type x = Rename<{ renameMe: string }, { "renameMe": "newName" }>
// x === { newName: string }
export type Rename<
  T,
  R extends {
    [K in keyof R]: K extends keyof T ? PropertyKey : "Error: key not in T";
  }
> = { [P in keyof T as P extends keyof R ? R[P] : P]: T[P] };
