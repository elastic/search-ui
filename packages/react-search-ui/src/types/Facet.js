import PropTypes from "prop-types";
import FacetType from "./FacetType";
import FacetValue from "./FacetValue";

export default PropTypes.shape({
  data: PropTypes.arrayOf(FacetValue).isRequired,
  // Name of the field this facet is associated with
  field: PropTypes.string.isRequired,
  type: FacetType.isRequired
});
