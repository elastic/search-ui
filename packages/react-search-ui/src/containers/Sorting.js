import PropTypes from "prop-types";
import { Component } from "react";
import { withSearch } from "..";
import { Sorting } from "@elastic/react-search-ui-views";
import SortList from "../types/SortList";

import { SortOption } from "../types";

function findSortOption(sortOptions, sortData) {
  if (sortData.indexOf("|||") === -1) {
    return sortOptions.find(
      option => JSON.stringify(option.value) === sortData
    );
  }

  const [value, direction] = sortData.split("|||");

  return sortOptions.find(
    option => option.value === value && option.direction === direction
  );
}

function formatValue(sortField, sortDirection, sortList) {
  if (sortList && sortList.length > 0) {
    return JSON.stringify(sortList);
  }
  return `${sortField}|||${sortDirection}`;
}

function formatSelectValues(sortOption) {
  if (Array.isArray(sortOption.value)) {
    // save value as string for comparison
    return JSON.stringify(sortOption.value);
  } else {
    return formatValue(sortOption.value, sortOption.direction);
  }
}

function formatSelectOption(sortOption) {
  return {
    label: sortOption.name,
    value: formatSelectValues(sortOption)
  };
}
export class SortingContainer extends Component {
  static propTypes = {
    // Props
    className: PropTypes.string,
    label: PropTypes.string,
    sortOptions: PropTypes.arrayOf(SortOption).isRequired,
    view: PropTypes.func,
    // State
    sortDirection: PropTypes.oneOf(["asc", "desc", ""]),
    sortField: PropTypes.string,
    sortList: SortList,
    // Actions
    setSort: PropTypes.func.isRequired
  };

  render() {
    const {
      className,
      label,
      setSort,
      sortDirection,
      sortField,
      sortList,
      sortOptions,
      view,
      ...rest
    } = this.props;

    const View = view || Sorting;

    return View({
      className,
      label,
      onChange: o => {
        const sortOption = findSortOption(sortOptions, o);
        setSort(sortOption.value, sortOption.direction);
      },
      options: sortOptions.map(formatSelectOption),
      value: formatValue(sortField, sortDirection, sortList),
      ...rest
    });
  }
}

export default withSearch(
  ({ sortDirection, sortField, sortList, setSort }) => ({
    sortDirection,
    sortField,
    sortList,
    setSort
  })
)(SortingContainer);
