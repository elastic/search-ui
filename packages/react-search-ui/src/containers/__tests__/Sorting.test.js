import React from "react";
import { shallow } from "enzyme";
import { SortingContainer } from "../Sorting";
import { SortOption } from "../../types";

const params = {
  results: [{}],
  searchTerm: "test",
  setSort: jest.fn(),
  sortDirection: "asc",
  sortField: "field",
  sortOptions: [
    SortOption.create({
      name: "name",
      value: "field",
      direction: "asc"
    }),
    SortOption.create({
      name: "name",
      value: "field",
      direction: "desc"
    })
  ]
};

const sortListParams = {
  ...params,
  sortList: [
    {
      field: "states",
      direction: "asc"
    },
    {
      field: "title",
      direction: "desc"
    }
  ],
  sortOptions: [
    SortOption.create({
      name: "name",
      value: "field",
      direction: "asc"
    }),
    SortOption.create({
      name: "name",
      value: "field",
      direction: "desc"
    }),
    {
      name: "multiple",
      value: [
        {
          field: "states",
          direction: "asc"
        },
        {
          field: "title",
          direction: "desc"
        }
      ]
    }
  ]
};

beforeEach(() => {
  params.setSort = jest.fn();
});

it("supports a render prop", () => {
  // eslint-disable-next-line react/prop-types
  const render = ({ value }) => {
    return <div>{value}</div>;
  };
  const wrapper = shallow(<SortingContainer {...params} view={render} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders when it doesn't have results or a search term", () => {
  const wrapper = shallow(
    <SortingContainer
      {...{
        ...params,
        searchTerm: "",
        results: []
      }}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("will call back when sort is changed in view", () => {
  let viewProps;

  shallow(<SortingContainer {...params} view={props => (viewProps = props)} />);

  const { onChange } = viewProps;
  onChange("field|||desc");

  const [sortField, sortDirection] = params.setSort.mock.calls[0];
  expect(sortField).toEqual("field");
  expect(sortDirection).toEqual("desc");
});

it("will call back when sort is changed in view with sortList", () => {
  let viewProps;

  shallow(
    <SortingContainer {...sortListParams} view={props => (viewProps = props)} />
  );

  const { onChange } = viewProps;
  onChange(
    '[{"field":"states","direction":"asc"},{"field":"title","direction":"desc"}]'
  );

  expect(sortListParams.setSort).toHaveBeenCalledWith(
    [
      { direction: "asc", field: "states" },
      { direction: "desc", field: "title" }
    ],
    undefined
  );
});

it("passes className through to the view", () => {
  let viewProps;
  const className = "test-class";
  shallow(
    <SortingContainer
      {...params}
      className={className}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps.className).toEqual(className);
});

it("passes data-foo through to the view", () => {
  let viewProps;
  const data = "bar";
  shallow(
    <SortingContainer
      {...params}
      data-foo={data}
      view={props => (viewProps = props)}
    />
  );
  expect(viewProps["data-foo"]).toEqual(data);
});
