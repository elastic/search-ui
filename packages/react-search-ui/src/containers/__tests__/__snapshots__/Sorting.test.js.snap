// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders when it doesn't have results or a search term 1`] = `
<div
  className="sui-sorting"
  results={Array []}
  searchTerm=""
>
  <StateManager
    className="sui-select"
    classNamePrefix="sui-select"
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "label": "name",
          "value": "field|||asc",
        },
        Object {
          "label": "name",
          "value": "field|||desc",
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
      }
    }
    value={
      Object {
        "label": "name",
        "value": "field|||asc",
      }
    }
  />
</div>
`;

exports[`supports a render prop 1`] = `
<div>
  field|||asc
</div>
`;
